 # Importing the Required Libraries and dataset

In [77]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import seaborn as sns
import matplotlib.pyplot as plt #for visualisation
from sklearn.model_selection import train_test_split


In [78]:
df = pd.read_csv("/home/sakshi/Downloads/cardataset/data.csv")
df.head(21)

Make       Model  Year             Engine Fuel Type  Engine HP  \
0    BMW  1 Series M  2011  premium unleaded (required)      335.0   
1    BMW    1 Series  2011  premium unleaded (required)      300.0   
2    BMW    1 Series  2011  premium unleaded (required)      300.0   
3    BMW    1 Series  2011  premium unleaded (required)      230.0   
4    BMW    1 Series  2011  premium unleaded (required)      230.0   
5    BMW    1 Series  2012  premium unleaded (required)      230.0   
6    BMW    1 Series  2012  premium unleaded (required)      300.0   
7    BMW    1 Series  2012  premium unleaded (required)      300.0   
8    BMW    1 Series  2012  premium unleaded (required)      230.0   
9    BMW    1 Series  2013  premium unleaded (required)      230.0   
10   BMW    1 Series  2013  premium unleaded (required)      300.0   
11   BMW    1 Series  2013  premium unleaded (required)      230.0   
12   BMW    1 Series  2013  premium unleaded (required)      300.0   
13   BMW    1 Series  2013  premium unleaded (required)      230.0   
14   BMW    1 Series  2013  premium unleaded (required)      230.0   
15   BMW    1 Series  2013  premium unleaded (required)      320.0   
16   BMW    1 Series  2013  premium unleaded (required)      320.0   
17  Audi         100  1992             regular unleaded      172.0   
18  Audi         100  1992             regular unleaded      172.0   
19  Audi         100  1992             regular unleaded      172.0   
20  Audi         100  1992             regular unleaded      172.0   

    Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
0                6.0            MANUAL   rear wheel drive              2.0   
1                6.0            MANUAL   rear wheel drive              2.0   
2                6.0            MANUAL   rear wheel drive              2.0   
3                6.0            MANUAL   rear wheel drive              2.0   
4                6.0            MANUAL   rear wheel drive              2.0   
5                6.0            MANUAL   rear wheel drive              2.0   
6                6.0            MANUAL   rear wheel drive              2.0   
7                6.0            MANUAL   rear wheel drive              2.0   
8                6.0            MANUAL   rear wheel drive              2.0   
9                6.0            MANUAL   rear wheel drive              2.0   
10               6.0            MANUAL   rear wheel drive              2.0   
11               6.0            MANUAL   rear wheel drive              2.0   
12               6.0            MANUAL   rear wheel drive              2.0   
13               6.0            MANUAL   rear wheel drive              2.0   
14               6.0            MANUAL   rear wheel drive              2.0   
15               6.0            MANUAL   rear wheel drive              2.0   
16               6.0            MANUAL   rear wheel drive              2.0   
17               6.0            MANUAL  front wheel drive              4.0   
18               6.0            MANUAL  front wheel drive              4.0   
19               6.0         AUTOMATIC    all wheel drive              4.0   
20               6.0            MANUAL  front wheel drive              4.0   

                          Market Category Vehicle Size Vehicle Style  \
0   Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                      Luxury,Performance      Compact   Convertible   
2                 Luxury,High-Performance      Compact         Coupe   
3                      Luxury,Performance      Compact         Coupe   
4                                  Luxury      Compact   Convertible   
5                      Luxury,Performance      Compact         Coupe   
6                      Luxury,Performance      Compact   Convertible   
7                 Luxury,High-Performance      Compact         Coupe   
8                                  Luxury      Compact   Convertible   
9                         

# Checking data
This is necessary to check which type of data are we dealing with like the data has null values or not whether the data is categorical whether string type of data is present ,checking duplicacy of data etc.

In [79]:
df.shape

(11914, 16)

In [80]:
df.describe()

Year    Engine HP  Engine Cylinders  Number of Doors  \
count  11914.000000  11845.00000      11884.000000     11908.000000   
mean    2010.384338    249.38607          5.628829         3.436093   
std        7.579740    109.19187          1.780559         0.881315   
min     1990.000000     55.00000          0.000000         2.000000   
25%     2007.000000    170.00000          4.000000         2.000000   
50%     2015.000000    227.00000          6.000000         4.000000   
75%     2016.000000    300.00000          6.000000         4.000000   
max     2017.000000   1001.00000         16.000000         4.000000   

        highway MPG      city mpg    Popularity          MSRP  
count  11914.000000  11914.000000  11914.000000  1.191400e+04  
mean      26.637485     19.733255   1554.911197  4.059474e+04  
std        8.863001      8.987798   1441.855347  6.010910e+04  
min       12.000000      7.000000      2.000000  2.000000e+03  
25%       22.000000     16.000000    549.000000  2.100000e+04  
50%       26.000000     18.000000   1385.000000  2.999500e+04  
75%       30.000000     22.000000   2009.000000  4.223125e+04  
max      354.000000    137.000000   5657.000000  2.065902e+06

In [81]:
df.isnull().any()

Make                 False
Model                False
Year                 False
Engine Fuel Type      True
Engine HP             True
Engine Cylinders      True
Transmission Type    False
Driven_Wheels        False
Number of Doors       True
Market Category       True
Vehicle Size         False
Vehicle Style        False
highway MPG          False
city mpg             False
Popularity           False
MSRP                 False
dtype: bool

In [82]:
df.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

In [83]:
for column in df.columns:
    if df[column].dtype == object:
        print(str(column) + ' : ' + str(df[column].unique()))
        print(df[column].value_counts())
        print("_________________________________________________________________")

Make : ['BMW' 'Audi' 'FIAT' 'Mercedes-Benz' 'Chrysler' 'Nissan' 'Volvo' 'Mazda'
 'Mitsubishi' 'Ferrari' 'Alfa Romeo' 'Toyota' 'McLaren' 'Maybach'
 'Pontiac' 'Porsche' 'Saab' 'GMC' 'Hyundai' 'Plymouth' 'Honda'
 'Oldsmobile' 'Suzuki' 'Ford' 'Cadillac' 'Kia' 'Bentley' 'Chevrolet'
 'Dodge' 'Lamborghini' 'Lincoln' 'Subaru' 'Volkswagen' 'Spyker' 'Buick'
 'Acura' 'Rolls-Royce' 'Maserati' 'Lexus' 'Aston Martin' 'Land Rover'
 'Lotus' 'Infiniti' 'Scion' 'Genesis' 'HUMMER' 'Tesla' 'Bugatti']
Chevrolet        1123
Ford              881
Volkswagen        809
Toyota            746
Dodge             626
Nissan            558
GMC               515
Honda             449
Mazda             423
Cadillac          397
Mercedes-Benz     353
Suzuki            351
BMW               334
Infiniti          330
Audi              328
Hyundai           303
Volvo             281
Subaru            256
Acura             252
Kia               231
Mitsubishi        213
Lexus             202
Buick             196
Chrysler

In [84]:
df.dtypes

Make                  object
Model                 object
Year                   int64
Engine Fuel Type      object
Engine HP            float64
Engine Cylinders     float64
Transmission Type     object
Driven_Wheels         object
Number of Doors      float64
Market Category       object
Vehicle Size          object
Vehicle Style         object
highway MPG            int64
city mpg               int64
Popularity             int64
MSRP                   int64
dtype: object

In [85]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
Make                 11914 non-null object
Model                11914 non-null object
Year                 11914 non-null int64
Engine Fuel Type     11911 non-null object
Engine HP            11845 non-null float64
Engine Cylinders     11884 non-null float64
Transmission Type    11914 non-null object
Driven_Wheels        11914 non-null object
Number of Doors      11908 non-null float64
Market Category      8172 non-null object
Vehicle Size         11914 non-null object
Vehicle Style        11914 non-null object
highway MPG          11914 non-null int64
city mpg             11914 non-null int64
Popularity           11914 non-null int64
MSRP                 11914 non-null int64
dtypes: float64(3), int64(5), object(8)
memory usage: 1.5+ MB


# Dropping Irrelevent columns

From above we can see that Market Category has may null values so we will directly drop this column and also column like Number of doors and Engine fuel type are not much related so we will also drop this two columns.


In [86]:
df = df.drop(['Engine Fuel Type', 'Number of Doors','Market Category'], axis=1)
df.head(5)

Make       Model  Year  Engine HP  Engine Cylinders Transmission Type  \
0  BMW  1 Series M  2011      335.0               6.0            MANUAL   
1  BMW    1 Series  2011      300.0               6.0            MANUAL   
2  BMW    1 Series  2011      300.0               6.0            MANUAL   
3  BMW    1 Series  2011      230.0               6.0            MANUAL   
4  BMW    1 Series  2011      230.0               6.0            MANUAL   

      Driven_Wheels Vehicle Size Vehicle Style  highway MPG  city mpg  \
0  rear wheel drive      Compact         Coupe           26        19   
1  rear wheel drive      Compact   Convertible           28        19   
2  rear wheel drive      Compact         Coupe           28        20   
3  rear wheel drive      Compact         Coupe           28        18   
4  rear wheel drive      Compact   Convertible           28        18   

   Popularity   MSRP  
0        3916  46135  
1        3916  40650  
2        3916  36350  
3        3916  29450  
4        3916  34500

In [87]:
df.shape

(11914, 13)

# Renaming Columns

In [88]:
df.rename(columns = {'Engine HP':'HP', 'Engine Cylinders':'Cylinders', 
                              'Transmission Type':'Transmission','Driven_Wheels': 'Drive Mode','highway MPG': 'MPG-H', 
                     'city mpg': 'MPG-C', 'MSRP': 'Price'}, inplace = True) 

In [89]:
df

Make       Model  Year     HP  Cylinders Transmission  \
0          BMW  1 Series M  2011  335.0        6.0       MANUAL   
1          BMW    1 Series  2011  300.0        6.0       MANUAL   
2          BMW    1 Series  2011  300.0        6.0       MANUAL   
3          BMW    1 Series  2011  230.0        6.0       MANUAL   
4          BMW    1 Series  2011  230.0        6.0       MANUAL   
...        ...         ...   ...    ...        ...          ...   
11909    Acura         ZDX  2012  300.0        6.0    AUTOMATIC   
11910    Acura         ZDX  2012  300.0        6.0    AUTOMATIC   
11911    Acura         ZDX  2012  300.0        6.0    AUTOMATIC   
11912    Acura         ZDX  2013  300.0        6.0    AUTOMATIC   
11913  Lincoln      Zephyr  2006  221.0        6.0    AUTOMATIC   

              Drive Mode Vehicle Size  Vehicle Style  MPG-H  MPG-C  \
0       rear wheel drive      Compact          Coupe     26     19   
1       rear wheel drive      Compact    Convertible     28     19   
2       rear wheel drive      Compact          Coupe     28     20   
3       rear wheel drive      Compact          Coupe     28     18   
4       rear wheel drive      Compact    Convertible     28     18   
...                  ...          ...            ...    ...    ...   
11909    all wheel drive      Midsize  4dr Hatchback     23     16   
11910    all wheel drive      Midsize  4dr Hatchback     23     16   
11911    all wheel drive      Midsize  4dr Hatchback     23     16   
11912    all wheel drive      Midsize  4dr Hatchback     23     16   
11913  front wheel drive      Midsize          Sedan     26     17   

       Popularity  Price  
0            3916  46135  
1            3916  40650  
2            3916  36350  
3            3916  29450  
4            3916  34500  
...           ...    ...  
11909         204  46120  
11910         204  56670  
11911         204  50620  
11912         204  50920  
11913          61  28995  

[11914 rows x 13 columns]

# Check for duplicates and null values and remove them

In [90]:
df.isnull().sum()

Make              0
Model             0
Year              0
HP               69
Cylinders        30
Transmission      0
Drive Mode        0
Vehicle Size      0
Vehicle Style     0
MPG-H             0
MPG-C             0
Popularity        0
Price             0
dtype: int64

In [91]:
#Columns HP and Cylinders have null values so we will remove them

In [97]:
df.dropna(inplace=True,axis=0)
df.shape

(11816, 13)

In [93]:
df.isnull().sum()

Make             0
Model            0
Year             0
HP               0
Cylinders        0
Transmission     0
Drive Mode       0
Vehicle Size     0
Vehicle Style    0
MPG-H            0
MPG-C            0
Popularity       0
Price            0
dtype: int64

In [94]:
now there are no null values


In [95]:
x=df[df.duplicated(keep='first')]
x

Make     Model  Year     HP  Cylinders Transmission  \
14        BMW  1 Series  2013  230.0        6.0       MANUAL   
18       Audi       100  1992  172.0        6.0       MANUAL   
20       Audi       100  1992  172.0        6.0       MANUAL   
24       Audi       100  1993  172.0        6.0       MANUAL   
25       Audi       100  1993  172.0        6.0       MANUAL   
...       ...       ...   ...    ...        ...          ...   
11481  Suzuki      X-90  1998   95.0        4.0       MANUAL   
11603   Volvo      XC60  2017  302.0        4.0    AUTOMATIC   
11604   Volvo      XC60  2017  240.0        4.0    AUTOMATIC   
11708  Suzuki       XL7  2008  252.0        6.0    AUTOMATIC   
11717  Suzuki       XL7  2008  252.0        6.0    AUTOMATIC   

              Drive Mode Vehicle Size Vehicle Style  MPG-H  MPG-C  Popularity  \
14      rear wheel drive      Compact         Coupe     28     19        3916   
18     front wheel drive      Midsize         Sedan     24     17        3105   
20     front wheel drive      Midsize         Sedan     24     17        3105   
24     front wheel drive      Midsize         Sedan     24     17        3105   
25     front wheel drive      Midsize         Sedan     24     17        3105   
...                  ...          ...           ...    ...    ...         ...   
11481   four wheel drive      Compact       2dr SUV     26     22         481   
11603    all wheel drive      Midsize       4dr SUV     29     20         870   
11604  front wheel drive      Midsize       4dr SUV     30     23         870   
11708    all wheel drive      Midsize       4dr SUV     22     15         481   
11717  front wheel drive      Midsize       4dr SUV     22     16         481   

       Price  
14     31500  
18      2000  
20      2000  
24      2000  
25      2000  
...      ...  
11481   2000  
11603  46350  
11604  40950  
11708  29149  
11717  27499  

[792 rows x 13 columns]

 From this we can see there are 792 duplicate rows so we will remove them

In [100]:
data=df.drop_duplicates(keep='first')

In [101]:
data.shape

(11024, 13)

In [102]:
data.head(21)

Make       Model  Year     HP  Cylinders Transmission         Drive Mode  \
0    BMW  1 Series M  2011  335.0        6.0       MANUAL   rear wheel drive   
1    BMW    1 Series  2011  300.0        6.0       MANUAL   rear wheel drive   
2    BMW    1 Series  2011  300.0        6.0       MANUAL   rear wheel drive   
3    BMW    1 Series  2011  230.0        6.0       MANUAL   rear wheel drive   
4    BMW    1 Series  2011  230.0        6.0       MANUAL   rear wheel drive   
5    BMW    1 Series  2012  230.0        6.0       MANUAL   rear wheel drive   
6    BMW    1 Series  2012  300.0        6.0       MANUAL   rear wheel drive   
7    BMW    1 Series  2012  300.0        6.0       MANUAL   rear wheel drive   
8    BMW    1 Series  2012  230.0        6.0       MANUAL   rear wheel drive   
9    BMW    1 Series  2013  230.0        6.0       MANUAL   rear wheel drive   
10   BMW    1 Series  2013  300.0        6.0       MANUAL   rear wheel drive   
11   BMW    1 Series  2013  230.0        6.0       MANUAL   rear wheel drive   
12   BMW    1 Series  2013  300.0        6.0       MANUAL   rear wheel drive   
13   BMW    1 Series  2013  230.0        6.0       MANUAL   rear wheel drive   
15   BMW    1 Series  2013  320.0        6.0       MANUAL   rear wheel drive   
16   BMW    1 Series  2013  320.0        6.0       MANUAL   rear wheel drive   
17  Audi         100  1992  172.0        6.0       MANUAL  front wheel drive   
19  Audi         100  1992  172.0        6.0    AUTOMATIC    all wheel drive   
21  Audi         100  1992  172.0        6.0       MANUAL    all wheel drive   
22  Audi         100  1993  172.0        6.0       MANUAL  front wheel drive   
23  Audi         100  1993  172.0        6.0    AUTOMATIC    all wheel drive   

   Vehicle Size Vehicle Style  MPG-H  MPG-C  Popularity  Price  
0       Compact         Coupe     26     19        3916  46135  
1       Compact   Convertible     28     19        3916  40650  
2       Compact         Coupe     28     20        3916  36350  
3       Compact         Coupe     28     18        3916  29450  
4       Compact   Convertible     28     18        3916  34500  
5       Compact         Coupe     28     18        3916  31200  
6       Compact   Convertible     26     17        3916  44100  
7       Compact         Coupe     28     20        3916  39300  
8       Compact   Convertible     28     18        3916  36900  
9       Compact   Convertible     27     18        3916  37200  
10      Compact         Coupe     28     20        3916  39600  
11      Compact         Coupe     28     19        3916  31500  
12      Compact   Convertible     28     19        3916  44400  
13      Compact   Convertible     28     19        3916  37200  
15      Compact   Convertible     25     18        3916  48250  
16      Compact         Coupe     28     20        3916  43550  
17      Midsize         Sedan     24     17        3105   2000  
19      Midsize         Wagon     20     16        3105   2000  
21      Midsize         Sedan     21     16        3105   2000  
22      Midsize         Sedan     24     17        3105   2000  
23      Midsize         Wagon     20     16        3105   2000